### ISSUE TITLE
Improved plain-text normalization for arXMLiv 08.2019

### ISSUE THREAD

**Issue by dginev** [KWARC/llamapun#31]
A list of improvements for the 2019 release of the various arXMLiv-derived resources:
- [x] math lexeme upgrades that expand ATOM and ARRAY (fixed in [latexml 1131](https://github.com/brucemiller/LaTeXML/pull/1131) )
- [x] refactor the parallel primitives for cleaner scaffolding of specific components 
- [x] have a path to word tokenization that does not require sentence tokenization (sensible when `PUNCT_.` is preserved). Done in 64063214761d9ae3ba5f839b147b8c1fb84464eb
- [x] add additional test for recovering malformed lexeme ends (see [latexml 723](https://github.com/brucemiller/LaTeXML/issues/723) for details)
- [x] consider including punctuation from the text modality, using the same lexemes as math punctuation (e.g. `PUNCT_,` and `OPEN_(`). Maybe follow the ASCII-mindful punctuation approach from BERT's [tokenization.py](https://github.com/google-research/bert/blob/master/tokenization.py)

This issue is now in a **testing period** over the 2018 data, in preparation for the 2019 release.

**Comment by dginev** [comment]
Another point worth mentioning w.r.t the GloVe+token models is capitalization of words. 

Currently the GloVe model models the capitalization, while the token model forces all words down to lowercase. That may be additional discrepancy that trips up Glove-leveraging models, so I'd have to make up my mind what the best practice ought to be in this case...

**Comment by dginev** [comment]
Since recently I have been exploring sentence-tokenized classification (for HAN document classification), it makes substantial sense that if the official token model contains the `ENDSENT` lexeme at the end of each sentence (or that lexeme is implied by the line-breaks, and can be used for additional classification), the capitalizations of the leading words can be used to separate the in-sentence word vectors from the lead-in word vectors, for the same word... Leading one to think keeping the capitalizations may be the way to go.

**Comment by dginev** [comment]
Recording the latest master (7f57ae30e33dc6e53d75e5dfba109722779d5a79) run over the 2018 version of the data via:
```
cargo run --release --example corpus_token_model /data/datasets/dataset-arXMLiv-08-2018/ /var/local/new_token_model.txt 
```

```
Token model finished in 11152s, gathered: 
1232184 documents;
85788441 paragraphs;
10005 discarded paragraphs (long words)
6873780841 words;
93052028 numeric literals;
480423348 formulas;
43669098 inline cites;
```

which is a clear increase over the [initial token model release](https://sigmathling.kwarc.info/resources/arxmliv-embeddings-082018/), by about 10% of data volume and seemingly better normalization footprint. Will add some extra measurements (including a GloVe model trained on it) for comparison.

**Comment by dginev** [comment]
One very positive sign is that the fallback normalization of `mathformula` which triggered on cases where the formula lexemes could not be detected changes as follows:

| version           | total formulas | `mathfromula` fallback  |
|:-----------------| -----------------:| ------------:|
| 2018 release |  455,929,408 | 6,981,315 |
| latest master |  480,423,348 | 163,113    |

That is a vast improvement over the lexematization while the formula count increased, the failure rate of lexeme normalization fell from 1.5% to 0.03%, and the new cases seem to almost always be occasions where the MathML contains an empty `<mi/>` something of the sort (low-level tex kerning code via math spaces with no content).

Looking pretty good!

**Comment by dginev** [comment]
Ok, published a 0.3.3 release and am preparing the embeddings with the final state of this issue. Good to go :+1:
